### PR TITLE
A Snake Game Application

### PR DESCRIPTION
**Pull request by tht-jxny**
This is to add a Snake Game App to wasp-os.
One can control the snake by swiping in the desired direction. 
Since only the necessary pixels are drawn again every tick, there is the problem that the first time the snake eats the food nothing seems to happen. I'd appreciate it if you could tell me how to fix this flaw.
The application comes with a Highscore. One can leave the app when the game is over.
And here is how it looks in action:
snake_game_in_action

### CODE REVIEW

File: wasp/apps/snake.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
+# SPDX-License-Identifier: LGPL-3.0-or-later
+# Copyright (C) 2020 Johannes Wache
+
+# 2-bit RLE, generated from res/snake_game.png, 856 bytes
```
Comment (daniel-thompson) [inline wasp/apps/snake.py:4]: As with previous PR, we need res/snake_game.png to be included as well as the resulting RLE code.

File: wasp/apps/snake.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+from apps.launcher import LauncherApp
+
+class SnakeGameApp():
+    NAME = 'Snake'
+    ICON = snake_game
+
+    def __init__(self):
+        self.running = True
+        self.snake = Snake()
+        self.foodLocation()
```
Comment (daniel-thompson) [inline wasp/apps/snake.py:76]: snake_case_please ;-)

File: wasp/apps/snake.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        wasp.watch.drawable.fill()
+        self._draw()
+        wasp.system.request_event(wasp.EventMask.TOUCH |
+                                  wasp.EventMask.SWIPE_UPDOWN |
+                                  wasp.EventMask.SWIPE_LEFTRIGHT)
+        wasp.system.request_tick(250)
+
+    def background(self):
+        """De-activate the application."""
+        pass
```
Comment (daniel-thompson) [inline wasp/apps/snake.py:91]: I think this function is redundant and can be removed.

File: wasp/apps/snake.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+    def sleep(self):
+        """Notify the application the device is about to sleep."""
+        return False
+
+    def wake(self):
+        """Notify the application the device is waking up."""
+        pass
```
Comment (daniel-thompson) [inline wasp/apps/snake.py:99]: Again... no need to implement the sleep/wake callbacks.

File: wasp/apps/snake.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    def __init__(self):
+        self.running = True
+        self.snake = Snake()
+        self.foodLocation()
+        self.highscore = 1
+        
+
+    def foreground(self):
+        """Activate the application."""
+        wasp.watch.drawable.fill()
```
Comment (daniel-thompson) [inline wasp/apps/snake.py:82]: If you have a _draw() method why put the fill here?

File: wasp/apps/snake.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+        if event[0] == wasp.EventType.UP:
+            self.snake.setDir(0,-15)
+        elif event[0] == wasp.EventType.DOWN:
+            self.snake.setDir(0,15)
+        elif event[0] == wasp.EventType.LEFT:
+            self.snake.setDir(-15,0)
+        elif event[0] == wasp.EventType.RIGHT:
+            self.snake.setDir(15,0)
+        else:
+            print("Error!")
```
Comment (daniel-thompson) [inline wasp/apps/snake.py:118]: Remove debugging code.

File: wasp/apps/snake.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+      else:
+        self.running = False
+        wasp.system.switch(LauncherApp())
```
Comment (daniel-thompson) [inline wasp/apps/snake.py:121]: No need to open code this. If you want to allow the swipe to work "as normal" when the game isn't running just `return True` from the swipe() funciton and the event will be passed to the system event handling.

File: wasp/apps/snake.py [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+    def tick(self, ticks):
+        """Notify the application that its periodic tick is due."""
+        self.update()
+        
+    
+    def foodLocation(self):
```
Comment (daniel-thompson) [inline wasp/apps/snake.py:128]: snake_case_please

File: wasp/apps/snake.py [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+# Based on https://www.youtube.com/watch?v=OMoVcohRgZA
+class Snake():
+    def __init__(self):
+        self.body = []
+        self.body.append([120,120])
+        self.xdir = 0
+        self.ydir = 0
+        self.len = 1
+    
+    def setDir(self,x,y):
```
Comment (daniel-thompson) [inline wasp/apps/snake.py:171]: snake_case

File: wasp/apps/snake.py [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+    
+    def eat(self,pos):
+        x = self.body[-1][0]
+        y = self.body[-1][1]
+        if (x == pos[0] and y == pos[1]):
+            self.grow()
+            return True
+        return False
+
+    def endGame(self):
```
Comment (daniel-thompson) [inline wasp/apps/snake.py:196]: snake_case

File: wasp/wasp.py [hunk 11]
```diff
@@ -121,6 +122,7 @@ def __init__(self):
                          (HeartApp, True),
                          (FlashlightApp, False),
                          (SettingsApp, False),
+                         (SnakeGameApp, False),
```
Comment (daniel-thompson) [inline wasp/wasp.py:125]: As before, let's leave this for users to add to main.py .

**Review by daniel-thompson — CHANGES_REQUESTED** [review]
Mostly fussy details in this review. I haven't tested as yet but I will keep your comment about draw problem during eating in mind.

BTW I also realized there's no docstrings in these apps and they are not integrated into the manual. I'd like to get the extra bonus apps (like Game of Life and your Snake game) into the docs and with screenshots. That will help people find them. It's fine to put it alongside other apps at the moment although I will probably move the optional apps into a user manual instead eventually.

**Comment by daniel-thompson** [comment]
What's the authorship of this code? I can see some of the code has come from another contributor but there's no Signed-off-by from the author. Is @wbjeffrey ok for this to be submitted (and able to provide a Signed-off-by: ).

**Comment by wbjeffrey** [comment]
@daniel-thompson  I'm not sure why I've been mentioned on this pull request. I don't have anything to do with this.

**Comment by daniel-thompson** [comment]
Ah... because I added this into the wrong PR... it should be in #74 where you are listed as the author of two the fibonaci clock patches.

**Comment by wbjeffrey** [comment]
@daniel-thompson That is really, really weird because I have nothing to do with that either. I don't know how my name is attached to that. I've not done anything other than the flash wake/sleep stuff.

**Comment by tht-jxny** [comment]
Regarding authorship: During my first commit in this branch, @wbjeffrey also caught my eye as an author. But I didn't care about it because I thought it wasn't that important.

**Comment by tht-jxny** [comment]
I could create a manual for the applications I created. @daniel-thompson . could you please start with the manual for your Game of Life application so I know what content is expected?

**Comment by daniel-thompson** [comment]
> I could create a manual for the applications I created. @daniel-thompson . could you please start with the manual for your Game of Life application so I know what content is expected?

Both the fibonacci clock and game of life are now included in the [application library](https://wasp-os.readthedocs.io/en/latest/apps.html) in the documentation. In both cases all that has been done is a simulator screenshot has been dropped into the res/ directory (play the game until the screen looks interesting, then press 's'), there has been markup text describing the app added at the top of the file and the app has been listed in doc/apps.rst .

**Comment by tht-jxny** [comment]
> > I could create a manual for the applications I created. @daniel-thompson . could you please start with the manual for your Game of Life application so I know what content is expected?
> 
> Both the fibonacci clock and game of life are now included in the [application library](https://wasp-os.readthedocs.io/en/latest/apps.html) in the documentation. In both cases all that has been done is a simulator screenshot has been dropped into the res/ directory (play the game until the screen looks interesting, then press 's'), there has been markup text describing the app added at the top of the file and the app has been listed in doc/apps.rst .

I'd really like to work on this but unfortunately, my fork is too far behind the current commit, so some files like `docs/apps.rst` are not yet available. This makes it difficult for me to make the desired changes, considering t that I'm actually a beginner when it comes to Git. Could you please give me some advice on how to address this problem?

**Comment by daniel-thompson** [comment]
> Could you please give me some advice on how to address this problem?

I would:

* Use `git branch` to make a back up of where you currently are in a seperate branch
* Use `get fetch origin` to get the latest code onto you rmachine (but not do anything with it)
* Use `git rebase origin/master` to replay your patches on top if the latest code... this is the bit where you need to pay attention because you may have to resolve merge conflicts. If you get conflicts read the prompts on-screen and take it slow. It basically means editing the conflicting file and looking for chevrons (`>>>>>>` or `<<<<<<` and editing that bit until it looks right again...  then continuing the rebase (you should only get conflicts on code you have previously edited so making it "look right again" isn't as hard as it sounds). If you do get stuck hopefully you can find a good blog post or video specifically about rebasing.
* Make some further changes, test, etc, etc, etc.
* Commit the changes as usual and then push them with `--force` flag in order to update the PR

So the two tricks are `git rebase` (which is tricky) and using `--force` on the git push when you're done (which is easy now you know about it but hard to figure out on your own).

This is also drilling some very useful git techniques! I'm a very heavy user of interactive rebases (`git rebase -i origin/master`) to update, edit and imporve the patches I have in my git repo before I push them. Interactive rebasing it a really good way to edit a pull request before force pushing an updated version.

**Comment by tht-jxny** [comment]
I hope everything works now. Thanks for your help @daniel-thompson. I also followed [this page](https://ardalis.com/syncing-a-fork-of-a-github-repository-with-upstream/).

**Comment by daniel-thompson** [comment]
Merged! Thanks.

**Pull request closed without merging**